### PR TITLE
Implemented trapdoors

### PR DESCRIPTION
**Pull request by DaPigGuy**
This PR implements trapdoors to Dragonfly

### CODE REVIEW

File: dragonfly/block/hashes.go [hunk 1]
```diff
@@ -31,8 +31,8 @@ const (
 	hashGranite
 	hashDiorite
 	hashAndesite
-	hasAndesite
```
Comment (TwistedAsylumMC) [inline dragonfly/block/hashes.go:34]: Might wanna add the h back to hash

File: dragonfly/block/trapdoor.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import (
+	"github.com/df-mc/dragonfly/dragonfly/block/wood"
+	"github.com/df-mc/dragonfly/dragonfly/entity/physics"
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+)
+
+//Trapdoor is a block that can be used as an openable 1x1 barrier
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:12]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	Wood wood.Wood
+	//Facing is the direction the trapdoor is facing
+	Facing world.Direction
+	//Open is whether or not the trapdoor is open
+	Open bool
+	//Top is whether the trapdoor occupies the top or bottom part of a block
+	Top bool
+}
+
+//AABB ...
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:27]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+)
+
+//Trapdoor is a block that can be used as an openable 1x1 barrier
+type Trapdoor struct {
+	noNBT
+
+	// Wood is the type of wood of the trapdoor. This field must have one of the values found in the material
+	// package.
+	Wood wood.Wood
+	//Facing is the direction the trapdoor is facing
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:19]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	Facing world.Direction
+	//Open is whether or not the trapdoor is open
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:21]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	Open bool
+	//Top is whether the trapdoor occupies the top or bottom part of a block
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:23]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+		return false
+	}
+	t.Facing = user.Facing().Opposite()
+	t.Top = (clickPos.Y() > 0.5 && face != world.FaceUp) || face == world.FaceDown
+
+	place(w, pos, t, user, ctx)
+	return placed(ctx)
+}
+
+//Activate ...
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:52]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	}
+}
+
+// CanDisplace ...
+func (t Trapdoor) CanDisplace(l world.Liquid) bool {
+	_, water := l.(Water)
+	return water
+}
+
+//EncodeItem ...
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:74]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		return -148, 0
+	case wood.Acacia():
+		return -145, 0
+	case wood.DarkOak():
+		return -147, 0
+	}
+	panic("invalid wood type")
+}
+
+//EncodeBlock ...
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:93]: Please add a space after the `//`.

File: dragonfly/block/trapdoor.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+	return BreakInfo{
+		Hardness:    3,
+		Harvestable: alwaysHarvestable,
+		Effective:   axeEffective,
+		Drops:       simpleDrops(item.NewStack(t, 1)),
+	}
+}
+
+// CanDisplace ...
+func (t Trapdoor) CanDisplace(l world.Liquid) bool {
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:69]: For the block to be waterloggable, the `world.LiquidDisplacer` interface has to be implemented, which also includes the `SideClosed` method. (Which I think should always return false in this case.)

File: dragonfly/block/trapdoor.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+	"github.com/df-mc/dragonfly/dragonfly/block/wood"
+	"github.com/df-mc/dragonfly/dragonfly/entity/physics"
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+)
+
+// Trapdoor is a block that can be used as an openable 1x1 barrier
+type Trapdoor struct {
```
Comment (Sandertv) [inline dragonfly/block/trapdoor.go:13]: Trapdoors don't diffuse light so you should implement the LightDiffuser interface and `return 0` for the diffusion level.

**Comment by Sandertv** [comment]
Thanks for the contribution!

**Pull request merged**